### PR TITLE
Refactor OAuth module documentation.

### PR DESCRIPTION
**Pull request by inconshreveable**
This commit refactors the OAuth module documentation to be consistent with our new module documentation format. It makes the following changes:

- Greatly expands documentation with additional information about how the module works including comprehensive examples for each connectivity method
- Fixes numerous errors about how the module works, mostly because the behavior of the module has changed over time
- Moved all of the per-provider information to the integration guides section.

One last note: I didn't finish the 'Errors' section. I'd like to come back to that and get this improved documentation out in front of the world without blocking on completing that section which requires a lot of in-depth spelunking time to pull together.

### CODE REVIEW

File: docs/cloud-edge/modules/oauth.mdx [hunk 1]
```diff
@@ -0,0 +1,700 @@
+# OAuth
```
[the remark this answers is not in the document]
Comment (inconshreveable) [reply docs/cloud-edge/modules/oauth.mdx:1]: TIL. what github preview functionality do you mean?

[the remark this answers is not in the document]
Comment (alawrenc) [reply docs/cloud-edge/modules/oauth.mdx:1]: using this RFC as an example https://github.com/ngrok-private/wiki/pull/37 , there's a link to the rendered view of the page

[the remark this answers is not in the document]
Comment (inconshreveable) [reply docs/cloud-edge/modules/oauth.mdx:1]: got it! for these we just use the vercel preview: https://ngrok-docs-git-alan-refactor-oauth-ngrok-dev.vercel.app/docs

image

File: docs/cloud-edge/modules/oauth.mdx [hunk 2]
```diff
@@ -0,0 +1,700 @@
...
+```
+
+###### **Customize requested authentication scopes**
+
+```
+ngrok http 80 \
+  --oauth github \
+  --oauth-client-id "{client id}" \
+  --oauth-client-secret "{client secret}" \
+  --scopes "repo" \
```
Comment (russorat) [inline docs/cloud-edge/modules/oauth.mdx:63]:
```suggestion
  --oauth-scope "repo" \
```

File: docs/cloud-edge/modules/oauth.mdx [hunk 3]
```diff
@@ -0,0 +1,700 @@
...
+  --scopes "user"
```
Comment (russorat) [inline docs/cloud-edge/modules/oauth.mdx:64]:
```suggestion
  --oauth-scope "user"
```

File: docs/cloud-edge/modules/oauth.mdx [hunk 4]
```diff
@@ -0,0 +1,700 @@
...
+```
+ngrok http 80 --oauth google
+```
+
+###### **Authenticate + Authorize by email and domain**
+
+```
+ngrok http 80 \
+  --oauth google \
+  --allow-email kate.libby@gmail.com \
```
Comment (russorat) [inline docs/cloud-edge/modules/oauth.mdx:43]:
```suggestion
  --oauth-allow-email kate.libby@gmail.com \
```

File: docs/cloud-edge/modules/oauth.mdx [hunk 5]
```diff
@@ -0,0 +1,700 @@
...
+  --allow-domain acme.org
```
Comment (russorat) [inline docs/cloud-edge/modules/oauth.mdx:44]:
```suggestion
  --oauth-allow-domain acme.org
```

File: docs/cloud-edge/modules/oauth.mdx [hunk 6]
```diff
@@ -0,0 +1,700 @@
...
+```bash
+ssh -R 443:localhost:80 connect.ngrok-agent.com http --oauth=google
+```
+
+###### **Authenticate + Authorize by email and domain**
+
+```bash
+ssh -R 443:localhost:80 connect.ngrok-agent.com http \
+  --oauth google \
+  --allow-email kate.libby@gmail.com \
```
Comment (russorat) [inline docs/cloud-edge/modules/oauth.mdx:133]:
```suggestion
  --oauth-allow-email kate.libby@gmail.com \
```

File: docs/cloud-edge/modules/oauth.mdx [hunk 7]
```diff
@@ -0,0 +1,700 @@
...
+  --allow-domain acme.org
```
Comment (russorat) [inline docs/cloud-edge/modules/oauth.mdx:134]:
```suggestion
  --oauth-allow-domain acme.org
```

File: docs/cloud-edge/modules/oauth.mdx [hunk 8]
```diff
@@ -0,0 +1,700 @@
...
+```
+
+###### **Customize requested authentication scopes**
+
+```bash
+ssh -R 443:localhost:80 connect.ngrok-agent.com http \
+  --oauth github \
+  --oauth-client-id "{client id}" \
+  --oauth-client-secret "{client secret}" \
+  --scopes "repo" \
```
Comment (russorat) [inline docs/cloud-edge/modules/oauth.mdx:153]:
```suggestion
  --oauth-scope "repo" \
```

File: docs/cloud-edge/modules/oauth.mdx [hunk 9]
```diff
@@ -0,0 +1,700 @@
...
+uniquely authenticated OAuth users who have accessed your endpoints within a
+month.
+
+OAuth MAU limits are enforced account-wide, they are _not_ enforced on a
+per-endpoint basis.
+
+An authenticated user with the same ID from the identity provider is counted as
+a single MAU even if they connect to multiple endpoints on your account.
+
+| Plan       | MAUs                            |
```
Comment (russorat) [inline docs/cloud-edge/modules/oauth.mdx:675]: should we document these limits here, or redirect users to the pricing page. i'm just thinking about future maintenence.

**Review by russorat — APPROVED** [review]
minor typos, but LGTM!

File: docs/cloud-edge/modules/oauth.mdx [hunk 10]
```diff
@@ -0,0 +1,700 @@
...
+
+### Upstream Headers {#upstream-headers}
+
+This module adds headers to the HTTP request sent to your upstream server with
+details about the OAuth user who has authenticated. These headers will not be
+set for `OPTIONS` requests if you enable options passthrough.
+
+| Header                           | Description                                                                   |
+| -------------------------------- | ----------------------------------------------------------------------------- |
+| `ngrok-auth-user-id`             | Provider-defined identified of the authorized user.                           |
```
Comment (rkolavo) [inline docs/cloud-edge/modules/oauth.mdx:614]:
```suggestion
| `ngrok-auth-user-id`             | Provider-defined identity of the authorized user.                           |
```

File: docs/cloud-edge/modules/oauth.mdx [hunk 11]
```diff
@@ -0,0 +1,700 @@
...
+
+An authenticated user with the same ID from the identity provider is counted as
+a single MAU even if they connect to multiple endpoints on your account.
+
+| Plan       | MAUs                            |
+| ---------- | ------------------------------- |
+| Free       | 5                               |
+| Personal   | 50                              |
+| Pro        | Unlimited, usaged-based-pricing |
+| Enterprise | Unlimited, usaged-based-pricing |
```
Comment (rkolavo) [inline docs/cloud-edge/modules/oauth.mdx:680]:
```suggestion
| Pro        | Unlimited, usage-based-pricing |
| Enterprise | Unlimited, usage-based-pricing |
```

File: docs/cloud-edge/modules/oauth.mdx (hunk 9 above)
Comment (inconshreveable) [reply docs/cloud-edge/modules/oauth.mdx]: probably. but right now there's no way to link to a specific line in the pricing table so our only option is to dump you on the pricing page at the top which feels really bad. i figure we can revisit this with the next rev of the pricing page and take this into consideration as part of the design. cc @samcrichard

**Pull request merged**